cmd/admin-speedtest.go

```diff
@@ -0,0 +1,107 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// speedTestDrivesService returns a channel of results of speed test
+// run on all drives
+func speedTestDrivesService(doneCh <-chan bool) (<-chan SpeedTestResultItem, error) {
+	// Build the list of storageAPI from global endpoints
+	var stDrives []StorageAPI
+	for _, ep := range globalEndpoints {
+		drive, err := newStorageAPI(ep)
+		if err != nil {
+			return nil, err
+		}
+		stDrives = append(stDrives, drive)
+	}
+
+	resCh := make(chan SpeedTestResultItem)
+	go func() {
+		// close the speed test result channel once we have
+		// sent all the results
+		defer close(resCh)
+
+		var wg sync.WaitGroup
+		for _, disk := range stDrives {
+			wg.Add(1)
+			go func(storage StorageAPI) {
+				defer wg.Done()
+				speedResItem := storage.SpeedTestDrive(minioMetaTmpBucket)
+				select {
+				case resCh <- speedResItem:
+					// ^Send the speedtest result on the channel.
+				case <-doneCh:
```

Comment:
did we miss `GlobalServiceDoneCh` here?

Reply:
@Praveenrajmani `speedTestDrivesReceiver` listens on `GlobalServiceDoneCh`. It closes the `doneCh` in the event it receives a signal on the `GlobalServiceDoneCh`.

Reply:
understood :+1: